File: app/src/main/java/it/unical/mat/embasp/base/Service.java
package it.unical.mat.embasp.base;

import java.util.List;

/**
 * Contains generic methods for ASP Solver execution .
 *
 * @see List
 */
public interface Service {

	/**
	 * Starts ASP solving Asyncronously on a subset of data and options.
	 *
	 * @param programs
	 *            a list of {@link InputProgram} used as data.
	 * @param options
	 *            a list of {@link OptionDescriptor} used as options.
	 * @param callback
	 *            interface used to interact with user
	 * @see Callback
	 */

	public void startAsync(Callback callback, List<InputProgram> programs, List<OptionDescriptor> options);

	/**
	 * Starts ASP solving Syncronously on a subset of data and options.
	 *
	 * @param programs
	 *            a list of {@link InputProgram} used as data.
	 * @param options
	 *            a list of {@link OptionDescriptor} used as options.
	 * @return {@link Output} element filled with results
	 */

	public Output startSync(List<InputProgram> programs, List<OptionDescriptor> options);

}

File: Python version/base/service.py
from abc import ABCMeta, abstractmethod


class Service(object):
    """Contains generic methods for ASP Solver execution."""

    __metaclass__ = ABCMeta

    @abstractmethod
    def start_async(self, callback, programs, options):
        """Starts ASP solving asynchronously on a subset of data and options.

        The parameter callback is an interface used to interact with
        user. The parameter programs represents a list of InputProgram
        used as data. The parameter options is a list of
        OptionDescriptor used as options.
        """
        pass

    @abstractmethod
    def start_sync(self, programs, options):
        """Starts ASP solving synchronously on a subset of data and options.

        The parameter programs is a list of InputProgram used as data.
        The parameter options is a list of OptionDescriptor used as
        options. The method return an Output element filled with
        results.
        """
        pass
